src/parser/tokenizer.rs

```diff
@@ -263,13 +264,36 @@ impl<'a> Tokenizer<'a> {
             => false,
             _ => !c.is_whitespace(),
         };
-        let (_, mut vec) = self.input.take_while(fun).map_err(to_token_error)?;
-        vec.insert(0, ch);
+        let (_, vec) = self.input.take_while(fun).map_err(to_token_error)?;
         Ok(Token::Key(pos, vec))
     }
 
-    pub fn next_token(&mut self) -> Result<Token, TokenError> {
-        let (pos, ch) = self.input.next_char().map_err(to_token_error)?;
+    pub fn next_token(&mut self) -> Result<Token<'a>, TokenError> {
+        let (pos, ch) = self.input.peek_char().map_err(to_token_error)?;
```

Comment:
I'm not sure about the logic, but are we changing behavior here? What are the situations in which we do `peek` instead of `next` as compared to the original?

Reply:
The old code was creating a new string. So it used `next`, saw the `$`, took the string after, and added it to char `$`. Now we do not want to compose a new string but return a slice that points to the original string, so we peek the next char, advance if needed (if its not `$`), and if it is `$` we take the entire slice (including the `$`).